test/snapshots/DeclarationBreaks.output

```diff
@@ -5,26 +5,114 @@ import Bar
 
 type Ok = String
 
-type OK2 :: Type
+type Ok2 :: Type
 type Ok2 = String
 
+type Ok3 = String
+type Ok4 = Int
+class Foo a where
+  foo :: a
+
+instance Foo Int where
+  foo = 12
+else instance Foo String where
+  foo = "foo"
+
 test :: Int
 test = 42
   where
-  a = 1
+  a =
+    1
+
   b = 2
 
   c :: Int
   c =
     let
       d :: Int
       d = 3
+
       e = 4
     in
       do
         let
-          f = 5
+          f =
+            5
 
           g :: Int
           g = 6
         g
+
+test2 Foo =
+  1
+test2 Bar =
+  2
+test2 Baz = 3
+
+data Wat = Wat1 | Wat2 | Wat3 | Wat4 | Wat5 | Wat6 | Wat7 | Wat8 | Wat9
+data Ohh = Ohh
+
+-- @format --width 40
+module DeclarationBreaks where
+
+import Foo
+import Bar
+
+type Ok = String
+
+type Ok2 :: Type
+type Ok2 = String
+
+type Ok3 = String
+type Ok4 = Int
+class Foo a where
```

Comment:
This seems like it would also benefit from the same rule and getting a newline before the class declaration

Reply:
I've pushed more fine-grained grouping that accounts for this.